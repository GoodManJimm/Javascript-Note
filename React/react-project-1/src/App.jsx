import { UserProfile } from "./Component/UserProfile"
import ReactDOM from 'react-dom/client'

export default function App()
{
    const callMe = () => 
    {
        console.log("Hello");
    }
    return (
    <div>
        <h1>Root Component</h1>
        <UserProfile age = {20} isLoggedIn = {true} 
        favoriteFoods = {[
        {
            name: "sushi",
            id: "Sushi"
        },
        {
            name: "pizza",
            id: "Pizza"
        }
        ]} callMe={callMe} username="bob"
        />
    </div>
    )
}