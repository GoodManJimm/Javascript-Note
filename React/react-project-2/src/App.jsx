import { useState } from 'react'
import ReactDOM from 'react-dom/client'
import { UserDetails } from './Components/UserDetails'
import LoginForm from './Components/LoginForm'
import RegisterForm from './Components/RegisterForm'

function App() {

  const [username,setUsername] = useState("")
  const [email,setEmail] = useState("")

  window.addEventListener('resize',(e) => 
    {
      console.log(window.innerHeight,window.innerWidth);
    })

  const [users,setUsers] = useState([ 
  {
    id: 1,
    username: "anson",
    email: "anson@gmail.com"
  },
  {
    id: 2,
    username: "bnson",
    email: "bnson@gmail.com"
  }])

  const [counter, setCounter] = useState(users.length + 1)
  console.log(counter);
  
  // const mockUsers = [
  //   {
  //     id: 1,
  //     username: "anson",
  //     email: "anson@gmail.com"
  //   },
  //   {
  //     id: 2,
  //     username: "bnson",
  //     email: "bnson@gmail.com"
  //   }
  // ]

  // const isLogin = false

  // return isLogin ? (<div>
  //   <h1>You are login</h1>
  // </div>) : (<div>
  //   <h1>You are not login</h1>
  // </div>)

  return (<div>
    {/* {mockUsers.map((user) => 
      {
        return (
        <UserDetails key={user.id} user={user}/>
      )
      })}
      <LoginForm/>
      <RegisterForm/> */}

      <div>
        <form onSubmit={(e) => {
          e.preventDefault();
          console.log(username,email);
          const newUser = 
          {
            id:counter,
            username : username,
            email : email
          }
          setCounter((counter) => counter + 1)
          setUsers((currentUsers) => [...currentUsers,newUser])
        }}>

          <div>
            <label htmlFor="username">Username: </label>
            <br />
            <input id='username' name='username' value={username} onChange={(e) => 
              {
                setUsername(e.target.value)
              }}/>
          </div>

          <div>
            <label htmlFor="email">Email: </label>
            <br />
            <input id='email' name='email' value={email} onChange={(e) => 
              {
                setEmail(e.target.value)
              }}/>
          </div>
          <button>Add User</button>
        </form>
      </div>
      {users.map((user) => <UserDetails key={user.id} user={user} setUsers={setUsers}/>)}


  </div>
  )
}

export default App
