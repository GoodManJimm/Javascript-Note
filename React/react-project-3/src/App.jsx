import { useState, useEffect } from "react";

export default function App(){
    
    const [counter, setCounter] = useState(0)
    const [sync,setSync] = useState(false)
    const [blogPostData,setBlogPostData] = useState({
        title: "",
        body: ""
    })


    useEffect(() => {
        console.log("rendering...");
        document.title = "react-project " + counter

    },[sync])


    useEffect(() => {

        const controller = new AbortController()
        console.log(controller.signal);

        async function fetchUsers()
        {
            try 
            {
                const response = await fetch("https://jsonplaceholder.typicode.com/users",
                    {signal: controller.signal}
                )
                const json = await response.json()
                console.log(json);
                console.log(controller.signal);
            }catch (err)
            {
                console.log(err);
            }
        }
        fetchUsers()
        return () => {
            controller.abort()
        }
        
    },[sync])


    return (
        <div>
            <form onSubmit={(e) => 
                {
                    e.preventDefault()
                    if (blogPostData.body || blogPostData.title)
                        {
                            fetch("https://jsonplaceholder.typicode.com/posts",
                            {
                                method: "POST",
                                body: JSON.stringify({
                                        userID : 1,
                                        title : blogPostData.title,
                                        body : blogPostData.body
                                    }),
                                    headers: {
                                        'Content-type': 'application/json; charset=UTF-8',
                                      },
                            }).then((response) => response.json())
                            .then((data) => {
                                    console.log("Success");
                                    console.log(data);
                                })
                                .catch((err) => {console.log(err);})
                        }
                }
            }>
                <label htmlFor="title">Title : </label>
                <input type="text" id="title" value={blogPostData.title} onChange={(e) => setBlogPostData((currentPost) => ({...currentPost,title : e.target.value}) )}/>
                <br />
                <label htmlFor="title">Body : </label>
                <input type="text" id="body" value={blogPostData.body} onChange={(e) => setBlogPostData((currentPost) => ({...currentPost,body : e.target.value}) )}/>
                <br />
                <button>Create New Post</button>
            </form>



            <div>{blogPostData.title}   {blogPostData.body}</div>
            <div>You clicked the button {counter}</div>
            <button onClick={() => setCounter((counter) => counter + 1)}>Click me</button>
            <br />
            <button onClick={() => setSync((current) => !current)}>Sync</button>
        </div>
    )
}