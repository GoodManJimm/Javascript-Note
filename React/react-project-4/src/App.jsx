import { useState } from 'react'
import LoginForm from './LoginForm'
import { UserContext } from './Context/UserContext'
import { PostContainer } from './Component/PostContainer'
import { PostContent } from './Component/PostContent'
import PostContentButton from './Component/PostContentButton'
import { useFetchUser } from './utils/hooks/useFetchUser'
import { useEffect } from 'react'
import { Outlet,Link, useNavigate } from 'react-router-dom'

function App() {
  
  const [toggle, setToggle] = useState(false)
  const [userData,setUserData] = useState()

  const {user,load, error} = useFetchUser(2)
  // console.log(user,load, error);

  const navigate = useNavigate()

  useEffect(() => 
    {
      if( !load && !error && user)
        {
          setUserData(user)
        }
    },[load,error,user,navigate])


  return (

  <UserContext.Provider value={{...userData,setUserData}}>

    <div>
      {load ? 'Loading...' : <PostContainer/>}
      <br />

      <button onClick={(e) => {
        setToggle((current) => !current)
        console.log(toggle);
      }}>Click me toggle</button>
      
      
      
      {/* toggle group */}
      
      {toggle && <LoginForm/>}
      {toggle && <PostContainer/>}
      {toggle && <PostContent/>}
      {toggle && <PostContentButton/>}



      <nav>
        <ul>
          <li>
            <Link to="/">Home</Link>
          </li>
          <li>
          <Link to="/users">Users</Link>
          </li>
          <li>
          <Link to="/blog-posts">Blog</Link>
          </li>
        </ul>
      </nav>

      <div>
        <label htmlFor="data">Enter Data</label>
        <input type="text" id='data' onChange={(e) => {
          console.log(e.target.value);
          if (e.target.value.length > 10)
            {
              navigate('/blog-posts',{
                state: {
                  posts: [{
                    id: 1,
                    title: "hello world",
                    content: "Welcome to my first post"
                  },{
                    id: 2,
                    title: "hello world",
                    content: "Welcome to my second post"
                  },],
                }
              })
            }
        }} />
      </div>

      <Outlet/>
    </div>
    </UserContext.Provider>
  )
}

export default App
