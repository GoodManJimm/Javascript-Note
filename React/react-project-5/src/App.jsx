import { useState } from 'react'
import LoginForm from './LoginForm'
import { UserContext } from './Context/UserContext'
import { PostContainer } from './Component/PostContainer'
import { PostContent } from './Component/PostContent'
import PostContentButton from './Component/PostContentButton'
import { useFetchUser } from './utils/hooks/useFetchUser'
import { useEffect } from 'react'
import { Outlet,Link, useNavigate } from 'react-router-dom'
import {UserDetails} from './Component/UserDetails'

function App({usersData}) {
  
  const [toggle, setToggle] = useState(true)

  const [userData,setUserData] = useState({
    id: 7000,
    username: 'johnny',
    email: 'johnny@gmail.com',
    name: 'Johnny',
    setUserData: () => {}
   })

  const [users,setUsers] = useState(usersData)

  const {user,load, error} = useFetchUser(2)
  // console.log(user,load, error);

  // const navigate = useNavigate()

  useEffect(() => 
    {
      if( !load && !error && user)
        {
          setUserData(user)
        }
    },[load,error,user])


  return (
  <> 

    {users.map((user) => (
      <UserDetails key={user.id} user={user} setUsers={setUsers} />
    ))}

    <UserContext.Provider value={{...userData,setUserData}}>

      <div>
        {/* {load ? 'Loading...' : <PostContainer/>} */}
        <br />

        <button onClick={(e) => {
          setToggle((current) => !current)
          console.log(toggle);
        }}>Click me toggle</button> 
        
        
        
        
        {/* {toggle && <LoginForm/>} */}
        {toggle && <PostContainer/>}
        {/* {toggle && <PostContent/>}
        {toggle && <PostContentButton/>} */}
        
      </div>
    </UserContext.Provider> 


    <nav>
      <ul>
        <li>
          <Link to="/">Home</Link>
        </li>
        <li>
        <Link to="/users">Users</Link>
        </li>
        <li>
        <Link to="/blog-posts">Blogs</Link>
        </li>
      </ul>
    </nav>
    {/*
    <div>
      <label htmlFor="data">Enter Data</label>
      <input type="text" id='data' onChange={(e) => {
        console.log(e.target.value);
        if (e.target.value.length > 10)
          {
            navigate('/blog-posts',{
              state: {
                posts: [{
                  id: 1,
                  title: "hello world",
                  content: "Welcome to my first post"
                },{
                  id: 2,
                  title: "hello world",
                  content: "Welcome to my second post"
                },],
              }
            })
          }
      }} />
    </div>
    {/* outlet decide where the chilren element appear */}
    <Outlet/> 
    
  </>
  )
}

export default App
