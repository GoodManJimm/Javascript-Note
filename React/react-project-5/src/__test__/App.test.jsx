import { describe, it, expect } from "vitest";
import { render ,screen, waitFor} from "@testing-library/react";
import App from "../App";
import {userEvent} from '@testing-library/user-event'
import { within } from "@testing-library/dom";
import { server } from "../__mocks__/msw/server";
import { http , HttpResponse, delay} from "msw";

describe('App', () => {
    it('should render users', () => {
        const { container } = render(<App usersData={[]}/>);
        expect(container).toMatchSnapshot();
    }),

    describe("When user is only 1", () => {
        it("should render save button" , async () => {
            render(<App usersData={[{
                id: 1,
                username: 'ha',
                email: 'ha@gmail.com'
              }]}/>)

            const editButton = screen.getByRole("button", {name:"Edit"})
            await userEvent.click(editButton)
            const saveButton = screen.getByRole("button", {name:"save"})
            expect(saveButton).toBeInTheDocument()
        }),
        it('should displau username & email input fields',async () => {
            render(<App usersData={[{
                id: 1,
                username: 'ha',
                email: 'ha@gmail.com'
              }]}/>)
            const editButton = screen.getByRole("button", {name:"Edit"})
            await userEvent.click(editButton)
            expect(screen.getByLabelText("Username:")).toBeInTheDocument()
            expect(screen.getByRole('textbox', {name : "email"})).toBeInTheDocument()
        })
    })

    describe("When there are 2 users", () => {
        it('should have two users' , () => {
            render(<App usersData={[
                {
                    id: 1,
                    username: 'ha',
                    email: 'ha@gmail.com'
                },
                {
                    id: 2,
                    username: 'ha2',
                    email: 'ha2@gmail.com'
                }
                ]}/>)

                expect(screen.getByText('ha')).toBeInTheDocument()
                expect(screen.getByText('ha2')).toBeInTheDocument()
        })
        it('should click edit button for 1st user and save button appear' ,async () => {
            render(<App usersData={[
            {
                id: 1,
                username: 'ha',
                email: 'ha@gmail.com'
            },
            {
                id: 2,
                username: 'ha2',
                email: 'ha2@gmail.com'
            }
            ]}/>)
            const userDetails = screen.getByTestId('user-details-1');
            expect(within(userDetails).getByText("ha")).toBeInTheDocument()
            expect(within(userDetails).queryByText("ha2")).toBeNull()
            
            const editButton = within(userDetails).getByRole("button", { name: "Edit" });
            await userEvent.click(editButton);
            expect(within(userDetails).getByRole("button", { name: "save" })).toBeInTheDocument();
            
        }),
        it('should click edit button for 1st user and save button appear' ,async () => {
            render(<App usersData={[
                {
                    id: 1,
                    username: 'ha',
                    email: 'ha@gmail.com'
                },
                {
                    id: 2,
                    username: 'ha2',
                    email: 'ha2@gmail.com'
                }
                ]}/>)
                const userDetails = screen.getByTestId('user-details-2');
                await userEvent.click(within(userDetails).getByRole('button',{name: 'Edit'}))
                await userEvent.type(within(userDetails).getByLabelText('Username:'),"123")
                await userEvent.click(within(userDetails).getByRole('button',{name: 'save'}))
                expect(within(userDetails).queryByLabelText("Username:")).toBeNull()
                expect(within(userDetails).getByText("ha2123")).toBeInTheDocument()
            })  
    }),

    describe('updating UserContext',() => {
        it("should update display name" , async () => {
            render(<App usersData={[]}/>)
            await userEvent.type(screen.getByLabelText("Update Name:"),"Jonathan The Dev")
            await userEvent.click(screen.getByRole('button',{name: "Update display name"}))
            expect(screen.getByText("Display Name:Jonathan The Dev")).toBeInTheDocument()
        })
    })


    describe('rendering context data', () => {
        it('should render correct email',async () => {

            server.use( http.get('https://jsonplaceholder.typicode.com/users/*', async ({params}) => {
                await delay(2000)
                return Response.json({
                    id: params.id,
                    username: 'josh',
                    name: 'josh',
                    email: 'josh@josh.com'
                })
            }))
            render(<App usersData={[]} />)
            expect(await screen.findByText("Email: josh@josh.com",{},{timeout:5000})).toBeInTheDocument()
            expect(screen.getByText("Email: josh@josh.com")).toBeInTheDocument();
            await waitFor(
                async () => {
                  expect(screen.getByText("Email: josh@josh.com")).toBeInTheDocument();
                },{timeout: 5000}
              );
        })
    })


});
