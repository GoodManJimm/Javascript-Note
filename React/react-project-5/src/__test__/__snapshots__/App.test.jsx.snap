// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should render users 1`] = `
<div>
  <div>
    <br />
    <button>
      Click me toggle
    </button>
    <div>
      <div>
        <span>
          Container
        </span>
      </div>
      <div>
        Display Name:
      </div>
      <div>
        ID: 
      </div>
      <div>
        Email: 
      </div>
      <div>
        Username: 
      </div>
      <div>
        <div>
          <h3>
            Post Content
          </h3>
          <br />
          <div>
            <div>
              <div>
                <span>
                  PostContentButton
                </span>
              </div>
              <br />
              <label
                for="UpdateName"
              >
                Update Name:
              </label>
              <input
                id="UpdateName"
                type="text"
                value=""
              />
              <button>
                Update display name
              </button>
            </div>
          </div>
        </div>
      </div>
      <br />
    </div>
  </div>
</div>
`;
